veles/secrets/gitlabpat/detector.go

```diff
@@ -0,0 +1,197 @@
+		token string
+	}
+
+	var versionedMatches, routableMatches, legacyMatches []match
+
+	// Collect routable versioned matches
+	contentStr := string(content)
+	for _, tokenMatchIndex := range reRoutableVersioned.FindAllStringSubmatchIndex(contentStr, -1) {
+		versionedMatches = append(versionedMatches, match{
+			start: tokenMatchIndex[0],
+			token: contentStr[tokenMatchIndex[0]:tokenMatchIndex[1]],
+		})
+	}
+
+	// Collect routable matches
+	for _, loc := range reRoutable.FindAllIndex(content, -1) {
+		routableMatches = append(routableMatches, match{
+			start: loc[0],
+			token: string(content[loc[0]:loc[1]]),
+		})
+	}
+
+	// Collect legacy matches
+	for _, loc := range reLegacy.FindAllIndex(content, -1) {
+		legacyMatches = append(legacyMatches, match{
+			start: loc[0],
+			token: string(content[loc[0]:loc[1]]),
+		})
+	}
+
+	pruned := make([]match, 0, len(versionedMatches)+len(routableMatches)+len(legacyMatches))
+
+	// Always keep versioned tokens
+	pruned = append(pruned, versionedMatches...)
+
+	// Keep routable tokens only if they're not contained in any versioned token
+	for _, routable := range routableMatches {
+		contained := false
+		for _, versioned := range versionedMatches {
+			if strings.Contains(versioned.token, routable.token) {
+				contained = true
+				break
+			}
+		}
+		if !contained {
+			pruned = append(pruned, routable)
+		}
+	}
+
+	// Keep legacy tokens only if they're not contained in any routable or versioned token
+	for _, legacy := range legacyMatches {
+		contained := false
+		// Check against versioned tokens
+		for _, versioned := range versionedMatches {
+			if strings.Contains(versioned.token, legacy.token) {
+				contained = true
+				break
+			}
+		}
+		// If not contained in versioned, check against routable
+		if !contained {
+			for _, routable := range routableMatches {
+				if strings.Contains(routable.token, legacy.token) {
+					contained = true
+					break
+				}
+			}
+		}
+		if !contained {
+			pruned = append(pruned, legacy)
+		}
+	}
+
+	// Filter out invalid versioned tokens based on CRC32 validation
+	finalMatches := make([]match, 0, len(pruned))
+	for _, m := range pruned {
+		if reRoutableVersioned.MatchString(m.token) {
+			submatch := reRoutableVersioned.FindStringSubmatch(m.token)
+			if len(submatch) > 0 {
+				var prefixIdx, payloadIdx, versionIdx, lengthIdx, crcIdx int
+				for i, name := range reRoutableVersioned.SubexpNames() {
+					if i == 0 {
+						continue
+					}
+					switch name {
+					case "prefix":
+						prefixIdx = i
+					case "payload":
+						payloadIdx = i
+					case "version":
+						versionIdx = i
+					case "length":
+						lengthIdx = i
+					case "crc":
+						crcIdx = i
+					}
+				}
+
+				if prefixIdx > 0 && payloadIdx > 0 && versionIdx > 0 && lengthIdx > 0 && crcIdx > 0 &&
+					isValidCRC32(submatch[prefixIdx], submatch[payloadIdx], submatch[versionIdx], submatch[lengthIdx], submatch[crcIdx]) {
+					finalMatches = append(finalMatches, m)
+				}
```

Comment:
Correct me if I was wrong. The current `reRoutableVersioned` all the groups are not empty, so the index should always be fixed? e.g. if matched it should always have `len(submatch)==6` and index will be 1-5? If so, the for loop is a bit overkilled, so it can also be  `len(submatch) ==6 && isValidCRC32(...){}`?